### PR TITLE
Added support for decoding .debug_pubtypes and .debug_pubnames sections

### LINKED ISSUE #201: Support for .debug_pubtypes and .debug_pubnames?
Hi, 

First of all, this is an incredibly useful tool that I've used to augment and automate many debug tasks. Thanks a lot for developing what I know of as the only Python-only framework for DWARF parsing. 

The one problem that I frequently face is the need to parse the entire ELF on start-up in order to generate indexing information for faster look-up based on three criteria - variable or function name, a type name (struct/enum etc.,), and address. I used to do this by manually walking the entire ELF and storing CU/DIE offsets as a function of names and addresses and types. This is quite time consuming especially when it comes to recursively parsing all the possible type definitions defined in the ELF.

I just recently learned that the DWARF format provides support for .debug_aranges, .debug_pubnames and .debug_pubtypes that pretty eliminates the need for the indexing that I am doing (as they already are the indexes that I need). Unfortunately, while .debug_aranges is supported, the other two (pubnames and pubtypes) aren't yet supported in pyelftools. 

I am wondering if there is a plan to add support for .debug_pubtypes and .debug_pubnames in the near future. If not, there is any other alternative to indexing the entire ELF for faster lookup?

Thanks,
Vijay.

### PR DESCRIPTION
**Pull request by rvijayc**
This pull request is meant to fix #201 . It adds support for parsing .debug_pubtypes and .debug_pubnames sections from an ELF file. 

There is an example file (examples/dwarf_pubnames_types.py) added that exercises the new features.

### CODE REVIEW

File: elftools/dwarf/namelut.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import os
+import collections
+from collections import OrderedDict
+from ..common.utils import struct_parse
+from bisect import bisect_right
+import math
+from ..construct import CString, Struct
+
+NameLUTEntry = collections.namedtuple('NameLUTEntry',
+        'cu_ofs die_ofs')
```
Comment (eliben) [inline elftools/dwarf/namelut.py]: Nit: don't wrap this

File: examples/dwarf_pubnames_types.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                print('Fetching the actual die for %s ...' % sym_name)
+                for cu in dwarfinfo.iter_CUs():
+                    if cu.cu_offset == entry.cu_ofs:
+                        for die in cu.iter_DIEs():
+                            if die.offset == entry.die_ofs:
+                                print(die)
+        
+            # dump all entries in .debug_pubnames section.
+            print('Dumping .debug_pubnames table ...')
+            print('-'*66)
```
Comment (eliben) [inline examples/dwarf_pubnames_types.py]: Here and everywhere else: spaces around operators 

```
'-' * 66
```

**Comment by rvijayc** [comment]
That's a good idea. I'll add the readelf support when I get a chance (along with other review comments), and add a new commit.

File: elftools/dwarf/namelut.py (hunk 1 above)
Comment (rvijayc) [reply elftools/dwarf/namelut.py]: I have resolved this in the latest commit.

File: examples/dwarf_pubnames_types.py (hunk 2 above)
Comment (rvijayc) [reply examples/dwarf_pubnames_types.py]: I have resolved this in the latest commit.

**Comment by rvijayc** [comment]
I've added the following in the latest commit:
- Added readelf tests for debug_pubnames and debug_pubtypes.
- fixed all review comments.
- Formatted doc strings.

**Pull request merged**